keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,242 @@
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
+}
+```
+
+### User Stories
+
+
+### Implementation Details/Notes/Constraints
+
+[CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) is used to find CSI driver name for in-tree volume plugins + its `VolumeHandle`. This CSI driver name is used as key in `CSINode.status.allocatable` list. The `VolumeHandle` is unique for each volume and will be used to de-duplicate volumes used by multiple pods on the same node.
+
+### Risks and Mitigations
+
+* This KEP depends on [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib). It can happen that CSI migration is redesigned / cancelled.
+  * Countermeasure: [CSI migration](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md) and this KEP should graduate together.
+
+* This KEP depends on [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) ability to handle in-line in-tree volumes. Scheduler will need to get CSI driver name + `VolumeHandle` from them to count them towards the limit. 
+
+## Design Details
+
+Existing feature gate `AttachVolumeLimit` will be re-used for implementation of this KEP. The feature is already beta and is enabled by default.
+
+### Test Plan
+
+* Run [scheduler benchmark](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-scheduling/scheduler_benchmarking.md) with matrix composed of:
+  * Predicates:
+    * All volume predicates enabled.
+    * Only deprecated `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` predicates enabled.
+    * Only `MaxCSIVolumeCountPred` predicate enabled.
+  * API objects:
+    * Both CSINode and Node containing `status.allocatable` for a volume plugin (to simulate kubelet during deprecation period).
+    * Only CSINode containing `status.allocatable` for a volume plugin (to simulate kubelet after deprecation period).
+    * Only Node containing `status.allocatable` for a volume plugin (to simulate old kubelet).
+  * Test results should be ideally the same as before the KEP.
+    * Any deviation needs to be approved by sig-scheduling.
+
+* Run e2e tests and kubelet version skew tests to check that scheduler picks the right values from CSINode or Node.
+
+* Add e2e test that runs pods with both in-tree volumes and CSI driver for the same storage backend and check that they share the same volume limits.
+
+### Graduation Criteria
+
+##### Alpha -> Beta Graduation
+
+N/A (`AttachVolumeLimit` feature is already beta).
+
+##### Beta -> GA Graduation
+
+It must graduate together with CSI migration.
```

Comment:
Do we want to consider adding a cache before GA to improve predicate performance?

Reply:
IMO, we need to have the implementation fast enough even in beta. It may turn out to be impossible to speed up the code enough after beta to reach GA requirements.

Reply:
may have missed it but what do we forsee being the slow part? The scheduler accessing CSINode API object?

Reply:
If I am not wrong- I think @msau42 meant caching of volumes *in-use* on a node. Currently the number of volumes in-use on a node is computed by iterating through running/in-flight pods and then adding number of volumes they are using. 

This can be cached and cache can be kept in-sync with decisions scheduler makes. Caching will potentially save some CPU cycles but IMO should not be targeted for this release. It can get fairly tricky pretty quickly.

Reply:
So for beta, our goal is same performance as the current implementation.  Before we go to GA, do we want to have a goal of improving performance by caching these in-use counts per node?

Reply:
added a line.